Fix -Wtautological-constant-out-of-range-compare warnings on 32-bit systems

hebasto opened the pull request:
On 32-bit system
```
$ uname -nrmo
odroid-hc1 4.14.187-odroidxu4 armv7l GNU/Linux
$ lsb_release -ds
Ubuntu 20.04.1 LTS
$ clang --version
clang version 10.0.0-4ubuntu1 
Target: armv7l-unknown-linux-gnueabihf
Thread model: posix
InstalledDir: /usr/bin
```

the clang fires the following warnings:
```
  CXX      policy/libbitcoin_common_a-feerate.o
policy/feerate.cpp:14:20: warning: result of comparison of constant 9223372036854775807 with expression of type 'size_t' (aka 'unsigned int') is always true [-Wtautological-constant-out-of-range-compare]
    assert(nBytes_ <= uint64_t(std::numeric_limits<int64_t>::max()));
           ~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/assert.h:93:27: note: expanded from macro 'assert'
     (static_cast <bool> (expr)                                         \
                          ^~~~
policy/feerate.cpp:25:20: warning: result of comparison of constant 9223372036854775807 with expression of type 'size_t' (aka 'unsigned int') is always true [-Wtautological-constant-out-of-range-compare]
    assert(nBytes_ <= uint64_t(std::numeric_limits<int64_t>::max()));
           ~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/assert.h:93:27: note: expanded from macro 'assert'
     (static_cast <bool> (expr)                                         \
                          ^~~~
2 warnings generated.
```

Fix is safe as https://github.com/bitcoin/bitcoin/blob/a57af897ec16976b28de05aa0b9c3f6a96d73ede/src/compat/assumptions.h#L55-L59

practicalswift commented:
Concept ACK

Please consider enabling `-Wtautological-constant-out-of-range-compare` too :)

hebasto commented:
> Please consider enabling `-Wtautological-constant-out-of-range-compare` too :)

Isn't it enabled by default?
https://clang.llvm.org/docs/DiagnosticsReference.html#wtautological-constant-out-of-range-compare

theStack commented:
Code Review ACK 7c30e9f6f2f18057366383cb2b0a54d99731041c

MarcoFalke commented on src/policy/feerate.cpp:
```diff
@@ -20,7 +19,6 @@ CFeeRate::CFeeRate(const CAmount& nFeePaid, size_t nBytes_)
 
 CAmount CFeeRate::GetFee(size_t nBytes_) const
 {
-    assert(nBytes_ <= uint64_t(std::numeric_limits<int64_t>::max()));
```
How is this `assert` save to remove for 64-bit?

hebasto replied:
Oh, my bad!

Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19735#issuecomment-834304120).

hebasto committed:
Fix -Wtautological-constant-out-of-range-compare warnings on 32-bit

It is safe, see assumptions.h:
"We assume size_t to be 32-bit or 64-bit."

hebasto committed:
Drop redundant `static_assert`s

It is safe, see assumptions.h:
"We assume size_t to be 32-bit or 64-bit."

hebasto commented:
Updated 7c30e9f6f2f18057366383cb2b0a54d99731041c -> 3759a5b3be7a88f265a282388af603b27e458ec0 ([pr19735.01](https://github.com/hebasto/bitcoin/commits/pr19735.01) -> [pr19735.02](https://github.com/hebasto/bitcoin/commits/pr19735.02)):
- rebased
- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/19735#discussion_r625671778)

hebasto commented:
The problem has been resolved in #21848 :tiger2:
